File: go/termites.go
// "Go to the Ant" §3.3 — Termite Nest Building (Kugler/Turvey; Parunak 1997), a
// faithful Go port of the authoritative Python reference (termites.py).
//
// Source: H. Van Dyke Parunak, "'Go to the Ant': Engineering Principles from
// Natural Multi-Agent Systems," Annals of Operations Research 75:69-101 (1997), §3.3.
//
// Tropical termites raise metre-high mounds — columns, arches, floors — with no
// chief engineer. Parunak's three local rules (§3.3):
//   1. Metabolize bodily waste, which contains pheromone. The waste IS the building
//      material.
//   2. Wander randomly, but prefer the direction of the strongest local pheromone.
//   3. Each step, decide stochastically whether to deposit the current load.
//      p(deposit) rises with the LOCAL pheromone density AND the amount carried. A
//      full termite drops even with no nearby deposit; a termite in a very high
//      local concentration drops even a small load.
// Because pheromone DECAYS, the freshest deposits (the centre of a growing pile)
// smell strongest, so piles climb upward into COLUMNS rather than spreading. No
// termite plans the mound.
//
// THE LENS (Go): termites are AGENTS (structs) stepping sequentially over one
// shared store (the Mound, holding two fields). The actor/goroutine framing lives
// in the naming (Termite agents, shared Mound), not yet in real parallelism —
// sequential stepping keeps the update deterministic and bit-identical to the
// other sequential ports.
//
// TWO fields (top-down 2D): mass = persistent structural mass (what you see);
// scent = decaying pheromone (what biases wandering). Emergence = scattered dabs
// self-concentrate into a HANDFUL of tall columns.
//
// Numbers differ from CPython only because this port uses the shared SplitMix64
// PRNG (identical across all language ports) rather than Python's Mersenne Twister;
// the emergence is the same.
package main

import (
	"flag"
	"fmt"
)

// ---- SplitMix64 PRNG (shared across all ports; standard library only) --------
// CROSS-PORT CONVENTION: these helpers intentionally do NOT reproduce CPython's
// random module; the goal is that all sequential ports agree with EACH OTHER.

type SplitMix64 struct{ state uint64 }

func (r *SplitMix64) next() uint64 {
	r.state += 0x9E3779B97F4A7C15
	z := r.state
	z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9
	z = (z ^ (z >> 27)) * 0x94D049BB133111EB
	return z ^ (z >> 31)
}

// random_float in [0,1): (next() >> 11) * 2^-53
func (r *SplitMix64) randomFloat() float64 {
	return float64(r.next()>>11) * (1.0 / 9007199254740992.0)
}

// randrange(n) -> int in [0,n): next() % n
func (r *SplitMix64) randrange(n int) int {
	return int(r.next() % uint64(n))
}

// DIRS: the 8 neighbours in THIS order (matches the reference exactly).
var DIRS = [8][2]int{
	{-1, -1}, {0, -1}, {1, -1}, {-1, 0}, {1, 0}, {-1, 1}, {0, 1}, {1, 1},
}

// ---- The shared store: Mound holds the two fields ----------------------------

type Mound struct {
	w, h  int
	rng   *SplitMix64
	mass  [][]float64 // persistent structure (viz)
	scent [][]float64 // dissipative pheromone (bias): diffuses AND decays
	buf   [][]float64 // double-buffer for the diffusion pass
	decay float64
	d     float64 // diffusion rate (Brownian spreading, §4.6)
}

func newGrid(h, w int) [][]float64 {
	g := make([][]float64, h)
	for y := range g {
		g[y] = make([]float64, w)
	}
	return g
}

func NewMound(w, h, seed int, decay float64) *Mound {
	return &Mound{
		w: w, h: h,
		rng:   &SplitMix64{state: uint64(seed)},
		mass:  newGrid(h, w),
		scent: newGrid(h, w),
		buf:   newGrid(h, w),
		decay: decay,
		d:     0.010,
	}
}

// fieldStep — the field law (§4.6 entropy leak): scent DIFFUSES (Brownian spreading,
// a local nearest-neighbour stencil) then EVAPORATES. Spreading gives each pile breadth
// — the substrate for column skirts and inter-column arches — while fresh cores still
// out-smell the spread. Double-buffered so every cell reads the OLD scent; draws NO rng,
// so cross-port bit-identity is preserved. new = old + D*(mean8 - old); then *= (1-decay).
func (m *Mound) fieldStep() {
	keep := 1.0 - m.decay
	for y := 0; y < m.h; y++ {
		for x := 0; x < m.w; x++ {
			acc := 0.0
			for _, d := range DIRS {
				nx := (x + d[0] + m.w) % m.w
				ny := (y + d[1] + m.h) % m.h
				acc += m.scent[ny][nx]
			}
			c := m.scent[y][x]
			m.buf[y][x] = (c + m.d*(acc/8.0-c)) * keep
		}
	}
	m.scent, m.buf = m.buf, m.scent
}

// columns: count distinct COLUMNS = toroidal local maxima above a fraction of the
// tallest peak. Returns (count, peak).
func (m *Mound) columns() (int, float64) {
	peak := 0.0
	for y := 0; y < m.h; y++ {
		for x := 0; x < m.w; x++ {
			if m.mass[y][x] > peak {
				peak = m.mass[y][x]
			}
		}
	}
	if peak <= 0 {
		return 0, 0.0
	}
	cut := peak * 0.15
	cnt := 0
	for y := 0; y < m.h; y++ {
		for x := 0; x < m.w; x++ {
			v := m.mass[y][x]
			if v < cut {
				continue
			}
			isMax := true
			for _, d := range DIRS {
				nx := (x + d[0] + m.w) % m.w
				ny := (y + d[1] + m.h) % m.h
				if v < m.mass[ny][nx] {
					isMax = false
					break
				}
			}
			if isMax {
				cnt++
			}
		}
	}
	return cnt, peak
}

// ---- The agent: a Termite that reads/writes the shared store -----------------

type Termite struct {
	m       *Mound
	x, y    int
	load    float64
	metab   float64
	maxload float64
}

func NewTermite(m *Mound, metab, maxload float64) *Termite {
	// RNG order matches the reference: x = randrange(w), then y = randrange(h).
	x := m.rng.randrange(m.w)
	y := m.rng.randrange(m.h)
	return &Termite{m: m, x: x, y: y, load: 0.0, metab: metab, maxload: maxload}
}

func (t *Termite) step() {
	m := t.m
	// rule 1: metabolize -> waste accumulates (capped at maxload).
	t.load = t.load + t.metab
	if t.load > t.maxload {
		t.load = t.maxload
	}
	// rule 2: wander, biased toward the strongest local scent over 8 toroidal
	// neighbours, weighted by (1 + scent*3).
	var wts [8]float64
	var nxs, nys [8]int
	tot := 0.0
	for i, d := range DIRS {
		nx := (t.x + d[0] + m.w) % m.w
		ny := (t.y + d[1] + m.h) % m.h
		w := 1.0 + m.scent[ny][nx]*3.0
		wts[i] = w
		nxs[i] = nx
		nys[i] = ny
		tot += w
	}
	r := m.rng.randomFloat() * tot
	for i := 0; i < 8; i++ {
		r -= wts[i]
		if r <= 0 {
			t.x = nxs[i]
			t.y = nys[i]
			break
		}
	}
	// rule 3: stochastic deposit — probability rises with local scent AND load; a
	// full termite always drops.
	local := m.scent[t.y][t.x]
	// OPERATIONALIZED: paper §3.3 gives NO formula, only "prob rises with local
	// density AND load". This deposit-probability formula is our operationalization.
	p := 0.01 + 0.55*(t.load/t.maxload) + 0.20*local
	if p > 1.0 {
		p = 1.0
	}
	// SHORT-CIRCUIT preserved from the reference: when load >= maxload the RNG is
	// NOT consumed (Python's `or` short-circuits), which keeps the stream in sync.
	if t.load >= t.maxload || m.rng.randomFloat() < p {
		m.mass[t.y][t.x] += t.load
		m.scent[t.y][t.x] += t.load
		t.load = 0.0
	}
}

// ---- Driver ------------------------------------------------------------------

func run(ticks, n, seed int, decay float64, verbose bool) (*Mound, []int) {
	metab, maxload := 0.4, 6.0
	mound := NewMound(58, 34, seed, decay)
	termites := make([]*Termite, n)
	for i := 0; i < n; i++ {
		termites[i] = NewTermite(mound, metab, maxload)
	}
	step := ticks / 12
	if step < 1 {
		step = 1
	}
	var hist []int
	for tk := 0; tk < ticks; tk++ {
		for _, tm := range termites {
			tm.step()
		}
		mound.fieldStep()
		if tk%step == 0 {
			c, _ := mound.columns()
			hist = append(hist, c)
		}
	}
	if verbose {
		render(mound)
		cnt, peak := mound.columns()
		fmt.Printf("\ndistinct columns (local maxima): %d | tallest column mass: %.0f\n", cnt, peak)
		fmt.Print("columns(t):")
		for _, c := range hist {
			fmt.Printf(" %d", c)
		}
		fmt.Println()
	}
	return mound, hist
}

func render(m *Mound) {
	peak := 0.0
	for y := 0; y < m.h; y++ {
		for x := 0; x < m.w; x++ {
			if m.mass[y][x] > peak {
				peak = m.mass[y][x]
			}
		}
	}
	if peak <= 0 {
		peak = 1.0
	}
	shades := " .:-=+*#%@"
	n := len(shades) - 1
	fmt.Print("\nTermite mound (top-down mass density — columns emerge as bright cores):\n\n")
	for y := 0; y < m.h; y++ {
		line := make([]byte, m.w)
		for x := 0; x < m.w; x++ {
			lvl := int(m.mass[y][x] / peak * float64(n))
			if lvl < 0 {
				lvl = 0
			}
			if lvl > n {
				lvl = n
			}
			line[x] = shades[lvl]
		}
		fmt.Println(string(line))
	}
}

func main() {
	ticks := flag.Int("ticks", 40000, "number of ticks")
	ants := flag.Int("ants", 70, "number of termites")
	decay := flag.Float64("decay", 0.02, "scent evaporation rate")
	seed := flag.Int("seed", 0, "PRNG seed")
	flag.Parse()

	run(*ticks, *ants, *seed, *decay, true)
}

File: js/termites.js
#!/usr/bin/env node
/*
 * "Go to the Ant" §3.3 — Termite Nest Building (Kugler/Turvey 1990), recreated
 * from the paper.
 *
 * Source: H. Van Dyke Parunak, "'Go to the Ant': Engineering Principles from
 * Natural Multi-Agent Systems," Annals of Operations Research 75:69-101 (1997).
 *
 * Tropical termites raise metre-high mounds — columns, arches, floors — with no
 * chief engineer. Parunak's three local rules (§3.3):
 *   1. Metabolize bodily waste, which contains pheromone. The waste IS the
 *      building material.
 *   2. Wander randomly, but prefer the direction of strongest local pheromone.
 *   3. Each step, decide stochastically whether to deposit the current load.
 *      p(deposit) rises with the LOCAL pheromone density AND the amount carried.
 *      A full termite drops even with no nearby deposit; a termite in a very high
 *      local concentration drops even a small load.
 * Because pheromone DECAYS, the freshest deposits (the centre of a growing pile)
 * smell strongest, so piles climb into COLUMNS rather than spreading. No termite
 * plans the mound. Emergence = scattered dabs self-concentrate into a handful of
 * tall columns.
 *
 * TWO local fields (top-down 2D): mass = persistent structural mass (what you
 * see); scent = decaying pheromone (what biases wandering).
 *
 * LENS (JavaScript): zero-dependency, standard-library only. Runs under node and
 * is portable to the browser visualizer — the web-native, accessible lens. All
 * 64-bit PRNG arithmetic uses BigInt (JS numbers are IEEE doubles); the field
 * math uses ordinary doubles, exactly as the Python reference does.
 */

'use strict';

// The 8 neighbours, in THIS exact order (matters for the weighted-random pick
// AND for the toroidal local-maximum test in columns()).
const DIRS = [[-1,-1],[0,-1],[1,-1],[-1,0],[1,0],[-1,1],[0,1],[1,1]];

// ---- PRNG: SplitMix64, identical across all ports (BigInt for u64 wrapping) ----
const MASK64 = (1n << 64n) - 1n;
class SplitMix64 {
  constructor(seed) { this.state = BigInt(seed) & MASK64; }
  next() {
    this.state = (this.state + 0x9E3779B97F4A7C15n) & MASK64;
    let z = this.state;
    z = ((z ^ (z >> 30n)) * 0xBF58476D1CE4E5B9n) & MASK64;
    z = ((z ^ (z >> 27n)) * 0x94D049BB133111EBn) & MASK64;
    return (z ^ (z >> 31n)) & MASK64;
  }
  // (next() >> 11) * 2^-53  ->  a double in [0, 1)
  random() { return Number(this.next() >> 11n) * (1.0 / 9007199254740992.0); }
  // randrange(n) -> int in [0, n)
  randrange(n) { return Number(this.next() % BigInt(n)); }
}

class Mound {
  constructor(w = 58, h = 34, seed = 0, decay = 0.02, D = 0.010) {
    this.w = w; this.h = h;
    this.rng = new SplitMix64(seed);
    this.mass = Array.from({ length: h }, () => new Float64Array(w));  // persistent structure (viz)
    this.scent = Array.from({ length: h }, () => new Float64Array(w)); // dissipative pheromone: diffuses AND decays
    this.buf = Array.from({ length: h }, () => new Float64Array(w));   // double-buffer for the diffusion pass
    this.decay = decay;
    this.D = D;                                                        // diffusion rate (Brownian spreading, §4.6)
  }

  // Field law (§4.6 entropy leak): scent DIFFUSES (Brownian spreading, a local nearest-
  // neighbour stencil) then EVAPORATES. Spreading gives each pile breadth — the substrate
  // for column skirts and inter-column arches. Double-buffered so every cell reads the OLD
  // scent; draws NO rng, so cross-port bit-identity is preserved.
  //   new = old + D*(mean8 - old);  then *= (1 - decay).
  fieldStep() {
    const keep = 1 - this.decay, D = this.D, w = this.w, h = this.h;
    for (let y = 0; y < h; y++) {
      const brow = this.buf[y];
      for (let x = 0; x < w; x++) {
        let acc = 0.0;
        for (const [dx, dy] of DIRS) {
          const nx = ((x + dx) % w + w) % w;
          const ny = ((y + dy) % h + h) % h;
          acc += this.scent[ny][nx];
        }
        const c = this.scent[y][x];
        brow[x] = (c + D * (acc / 8.0 - c)) * keep;
      }
    }
    const t = this.scent; this.scent = this.buf; this.buf = t;
  }

  // Count distinct COLUMNS = toroidal local maxima with mass above a fraction of
  // the tallest peak.
  columns() {
    let peak = 0.0;
    for (let y = 0; y < this.h; y++)
      for (let x = 0; x < this.w; x++)
        if (this.mass[y][x] > peak) peak = this.mass[y][x];
    if (peak <= 0) return [0, 0.0];
    const cut = peak * 0.15;
    let cnt = 0;
    for (let y = 0; y < this.h; y++) {
      for (let x = 0; x < this.w; x++) {
        const v = this.mass[y][x];
        if (v < cut) continue;
        let isMax = true;
        for (const [dx, dy] of DIRS) {
          const nx = ((x + dx) % this.w + this.w) % this.w;
          const ny = ((y + dy) % this.h + this.h) % this.h;
          if (v < this.mass[ny][nx]) { isMax = false; break; }
        }
        if (isMax) cnt += 1;
      }
    }
    return [cnt, peak];
  }
}

class Termite {
  constructor(m, metab = 0.4, maxload = 6.0) {
    this.m = m;
    // RNG order matches Python: randrange(w) then randrange(h), per termite.
    this.x = m.rng.randrange(m.w);
    this.y = m.rng.randrange(m.h);
    this.load = 0.0; this.metab = metab; this.maxload = maxload;
  }

  step() {
    const m = this.m;
    // rule 1: metabolize -> waste accumulates
    this.load = Math.min(this.maxload, this.load + this.metab);
    // rule 2: wander, biased toward the strongest local scent
    const wts = new Array(8);
    let tot = 0.0;
    for (let i = 0; i < 8; i++) {
      const nx = ((this.x + DIRS[i][0]) % m.w + m.w) % m.w;
      const ny = ((this.y + DIRS[i][1]) % m.h + m.h) % m.h;
      const w = 1.0 + m.scent[ny][nx] * 3.0;
      wts[i] = [w, nx, ny];
      tot += w;
    }
    let r = m.rng.random() * tot;
    for (let i = 0; i < 8; i++) {
      r -= wts[i][0];
      if (r <= 0) { this.x = wts[i][1]; this.y = wts[i][2]; break; }
    }
    // rule 3: stochastic deposit — rises with local scent AND load; full termite
    // always drops.
    const local = m.scent[this.y][this.x];
    // OPERATIONALIZED: paper §3.3 gives NO formula, only "prob rises with local
    // density AND load". This deposit-probability formula operationalizes that.
    const p = Math.min(1.0, 0.01 + 0.55 * (this.load / this.maxload) + 0.20 * local);
    // NB: short-circuit matches Python — when load>=maxload, random() is NOT
    // consumed, so the RNG stream stays bit-identical across ports.
    if (this.load >= this.maxload || m.rng.random() < p) {
      m.mass[this.y][this.x] += this.load;
      m.scent[this.y][this.x] += this.load;
      this.load = 0.0;
    }
  }
}

function run(ticks, n, seed, decay, verbose) {
  const mound = new Mound(58, 34, seed, decay);
  const termites = Array.from({ length: n }, () => new Termite(mound));
  const hist = [];
  const stride = Math.max(1, Math.floor(ticks / 12));
  for (let t = 0; t < ticks; t++) {
    for (const tm of termites) tm.step();
    mound.fieldStep();
    if (t % stride === 0) hist.push(mound.columns()[0]);
  }
  if (verbose) {
    render(mound);
    const [cnt, peak] = mound.columns();
    console.log(`\ndistinct columns (local maxima): ${cnt} | tallest column mass: ${peak.toFixed(0)}`);
    console.log('columns(t): ' + hist.join(' '));
  }
  return { mound, hist };
}

function render(mound) {
  let peak = 0.0;
  for (let y = 0; y < mound.h; y++)
    for (let x = 0; x < mound.w; x++)
      if (mound.mass[y][x] > peak) peak = mound.mass[y][x];
  if (peak === 0.0) peak = 1.0;
  const shades = ' .:-=+*#%@';
  console.log('\nTermite mound (top-down mass density — columns emerge as bright cores):\n');
  for (let y = 0; y < mound.h; y++) {
    let line = '';
    for (let x = 0; x < mound.w; x++) {
      let lvl = Math.floor((mound.mass[y][x] / peak) * (shades.length - 1));
      lvl = Math.max(0, Math.min(shades.length - 1, lvl));
      line += shades[lvl];
    }
    console.log(line);
  }
}

function main() {
  const args = process.argv.slice(2);
  const opts = { ticks: 40000, termites: 70, decay: 0.02, seed: 0 };
  for (let i = 0; i < args.length; i++) {
    const a = args[i];
    if (a === '--ticks') opts.ticks = parseInt(args[++i], 10);
    else if (a === '--termites' || a === '--ants') opts.termites = parseInt(args[++i], 10);
    else if (a === '--decay') opts.decay = parseFloat(args[++i]);
    else if (a === '--seed') opts.seed = parseInt(args[++i], 10);
  }
  run(opts.ticks, opts.termites, opts.seed, opts.decay, true);
}

main();
